drivers/flash/flash_stm32l5x.c

```diff
@@ -178,16 +180,27 @@ static int write_dword(const struct device *dev, off_t offset, uint64_t val)
 		return rc;
 	}
 
-	/* Check if this double word is erased and value isn't 0.
+	/* Check if this double/quad word is erased and value isn't 0.
 	 *
-	 * It is allowed to write only zeros over an already written dword
+	 * It is allowed to write only zeros over an already written dword / qword
 	 * See 6.3.7 in STM32L5 reference manual.
 	 * See 7.3.7 in STM32U5 reference manual.
+	 * See 7.3.5 in STM32H5 reference manual.
 	 */
-	if ((flash[0] != 0xFFFFFFFFUL ||
-	     flash[1] != 0xFFFFFFFFUL) && val != 0UL) {
-		LOG_ERR("Word at offs %ld not erased", (long)offset);
-		return -EIO;
+	full_zero = true;
+	for (i = 0; i < FLASH_STM32_WRITE_BLOCK_SIZE; i++) {
+		if (buff[i] != 0) {
+			full_zero = false;
+			break;
+		}
+	}
+	if (full_zero) {
```

Comment:
If I'm not mistaking, you're reverting the logic here.
It would be clearer to check:
- first: is the target zone not erased
- second: if not erased, verify that value is full_zeroes, error out if not the case.

Reply:
Indeed it is reversed, but this is on purpose: the flash access is presumably much slower than a RAM access. So it is faster to check if the value is zero, and only then check the flash in the rare event that we write 0x00(...)00. In most cases, we can completely skip reading the flash. I can add a comment to make it more clear.

Reply:
ok, then you should test `if (!full_zero) {`

Reply:
oh sure, sorry for this bug